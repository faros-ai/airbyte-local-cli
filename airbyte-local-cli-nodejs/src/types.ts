import {Dictionary} from 'ts-essentials';

export interface AirbyteCliContext {
  tmpDir?: string;
}

// Cli options provided by the user
export interface CliOptions {
  // source and destination config
  configFile?: string;
  wizard?: object;

  // source
  srcImage: string;
  srcConfig?: Record<string, any>;
  srcOutputFile?: string;
  srcCheckConnection?: boolean;
  srcOnly?: boolean;
  srcPull?: boolean;

  // destination
  dstImage: string;
  dstConfig?: Record<string, any>;
  dstUseHostNetwork?: boolean;
  dstOnly?: string;
  dstPull?: boolean;

  // general connector settings
  connectionName?: string;
  stateFile?: string;
  fullRefresh?: boolean;
  rawMessages?: boolean;
  keepContainers?: boolean;
  logLevel?: string;

  // logging
  debug?: boolean;

  // deprecated
  checkConnection?: boolean; // use srcCheckConnection instead
  state?: string; // use stateFile instead
}

// Airbyte connector source or destination config
export interface AirbyteConfig {
  image: string;
  config?: object;
  catalog?: AirbyteConfiguredCatalog;
  dockerOptions?: AirbyteConfigDockerOptions;
}
export interface AirbyteConfigDockerOptions {
  maxMemory?: number; // unit: MB
  maxCpus?: number; // unit: CPU
  maxLogSize?: string; // default: 10m (10MB)
  additionalOptions?: any;
}
export enum AirbyteConfigInputType {
  FILE = 'file',
  OPTION = 'option',
}

// Config that are needed for running this Airbyte Cli
export interface FarosConfig {
  src?: AirbyteConfig;
  dst?: AirbyteConfig;
  srcOutputFile: string | undefined; // if srcOnly is true
  srcInputFile: string | undefined; // if dstOnly is true
  srcCheckConnection: boolean;
  dstUseHostNetwork: boolean;
  srcPull: boolean;
  dstPull: boolean;
  connectionName: string | undefined;
  stateFile: string | undefined;
  fullRefresh: boolean;
  rawMessages: boolean;
  keepContainers: boolean;
  logLevel: string;
  debug: boolean;

  // internal use
  dstStreamPrefix?: string;
}

/**
 * Copy types from faros-ai/airtype-connectors
 * https://github.com/faros-ai/airbyte-connectors/blob/main/faros-airbyte-cdk/src/protocol.ts
 */
export enum AirbyteConnectionStatus {
  SUCCEEDED = 'SUCCEEDED',
  FAILED = 'FAILED',
}
export enum AirbyteMessageType {
  CATALOG = 'CATALOG',
  CONNECTION_STATUS = 'CONNECTION_STATUS',
  LOG = 'LOG',
  RECORD = 'RECORD',
  SPEC = 'SPEC',
  STATE = 'STATE',
  TRACE = 'TRACE',
}
export interface AirbyteMessage {
  readonly type: AirbyteMessageType;
}
export declare class AirbyteConnectionStatusMessage implements AirbyteMessage {
  readonly connectionStatus: {
    status: AirbyteConnectionStatus;
    message?: string;
  };
  readonly type: AirbyteMessageType;
  constructor(connectionStatus: {status: AirbyteConnectionStatus; message?: string});
}
export enum SyncMode {
  FULL_REFRESH = 'full_refresh',
  INCREMENTAL = 'incremental',
}
export enum DestinationSyncMode {
  APPEND = 'append',
  OVERWRITE = 'overwrite',
  APPEND_DEDUP = 'append_dedup',
}
export interface AirbyteStream {
  name: string;
  json_schema?: Dictionary<any>;
  supported_sync_modes?: SyncMode[];
  source_defined_cursor?: boolean;
  default_cursor_field?: string[];
  source_defined_primary_key?: string[][];
  namespace?: string;
}
export interface AirbyteCatalog {
  streams: AirbyteStream[];
}
export declare class AirbyteCatalogMessage implements AirbyteMessage {
  readonly catalog: AirbyteCatalog;
  readonly type: AirbyteMessageType;
  constructor(catalog: AirbyteCatalog);
}
export interface AirbyteConfiguredStream {
  stream: AirbyteStream;
  sync_mode: SyncMode;
  cursor_field?: string[];
  destination_sync_mode?: DestinationSyncMode;
  primary_key?: string[][];
  disabled?: boolean;
}
export interface AirbyteConfiguredCatalog {
  streams: AirbyteConfiguredStream[];
}
