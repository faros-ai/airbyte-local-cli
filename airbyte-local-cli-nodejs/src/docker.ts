import {writeFileSync} from 'node:fs';
import {Writable} from 'node:stream';

import Docker from 'dockerode';

import {AirbyteConnectionStatus, AirbyteConnectionStatusMessage, AirbyteMessageType, FarosConfig} from './types';
import {DEFAULT_STATE_FILE, logger, SRC_CATALOG_FILENAME, SRC_CONFIG_FILENAME} from './utils';

// Constants
const DEFAULT_MAX_LOG_SIZE = '10m';

// Create a new Docker instance
let _docker = new Docker();

// For testing purposes
export function setDocker(docker: Docker): void {
  _docker = docker;
}

export async function checkDockerInstalled(): Promise<void> {
  try {
    await _docker.version();
    logger.debug('Docker is installed and running.');
  } catch (error: any) {
    logger.error('Docker is not installed or running.');
    throw error;
  }
}

export async function pullDockerImage(image: string): Promise<void> {
  logger.info(`Pulling docker image: ${image}`);

  try {
    const stream = await _docker.pull(image);
    await new Promise((resolve, reject) => {
      _docker.modem.followProgress(stream, (err, res) => (err ? reject(err) : resolve(res)));
    });
    logger.info(`Docker image pulled: ${image}`);
  } catch (error: any) {
    logger.error(`Failed to pull docker image: ${image}`);
    throw error;
  }
}

/**
 * Spinning up a docker container to check the source connection.
 * `docker run --rm -v "$tempdir:/configs" $src_docker_options "$src_docker_image"
 *      check --config "/configs/$src_config_filename"`
 *
 * Sample output from the docker container:
 * {"connectionStatus":{"status":"SUCCEEDED"},"type":"CONNECTION_STATUS"}
 * {"connectionStatus":{"status":"FAILED","message":"Faros API key was not provided"},"type":"CONNECTION_STATUS"}
 */
export async function checkSrcConnection(tmpDir: string, image: string, srcConfigFile?: string): Promise<void> {
  logger.info('Validating connection to source...');

  if (!image) {
    throw new Error('Source image is missing.');
  }

  try {
    const cfgFile = srcConfigFile ?? SRC_CONFIG_FILENAME;
    const command = ['check', '--config', `/configs/${cfgFile}`];
    const createOptions: Docker.ContainerCreateOptions = {
      HostConfig: {
        Binds: [`${tmpDir}:/configs`],
        AutoRemove: true,
      },
      platform: 'linux/amd64',
    };

    // create a writable stream to capture the output
    let data = '';
    const outputStream = new Writable({
      write(chunk, _encoding, callback) {
        data += chunk.toString();
        callback();
      },
    });

    // docker run
    const res = await _docker.run(image, command, outputStream, createOptions);

    // capture connection status from the output
    let status: AirbyteConnectionStatusMessage | undefined;
    data.split('\n').forEach((line) => {
      if (line.includes(AirbyteMessageType.CONNECTION_STATUS)) {
        status = JSON.parse(line) as AirbyteConnectionStatusMessage;
      }
    });
    if (
      status?.type === AirbyteMessageType.CONNECTION_STATUS &&
      status?.connectionStatus.status === AirbyteConnectionStatus.SUCCEEDED &&
      res[0].StatusCode === 0
    ) {
      logger.info('Source connection is valid.');
    } else {
      throw new Error(status?.connectionStatus.message);
    }
  } catch (error: any) {
    throw new Error(`Failed to validate source connection: ${error.message ?? JSON.stringify(error)}.`);
  }
}

/**
 * Spinning up a docker container to run source airbyte connector.
 * Platform is set to 'linux/amd64' as we only publish airbyte connectors images for linux/amd64.
 *
 * Docker cli command:
 *  docker run --name $src_container_name --init \
 *    -v "$tempdir:/configs" \
 *    $max_memory $max_cpus --log-opt max-size="$max_log_size" \
 *    --env LOG_LEVEL="$log_level" \
 *    $src_docker_options
 *    --cidfile="$tempPrefix-src_cid" \
 *    -a stdout -a stderr \
 *    "$src_docker_image" \
 *    read \
 *    --config "/configs/$src_config_filename" \
 *    --catalog "/configs/$src_catalog_filename" \
 *    --state "/configs/$src_state_filename"
 *
 * @argument command - for testing purposes only
 */
export async function runSrcSync(tmpDir: string, config: FarosConfig, command?: string[]): Promise<string> {
  logger.info('Running source connector...');

  if (!config.src?.image) {
    throw new Error('Source image is missing.');
  }

  try {
    const timestamp = Date.now();
    const srcContainerName = `airbyte-local-src-${timestamp}`;
    const cmd = command ?? [
      'read',
      '--config',
      `/configs/${SRC_CONFIG_FILENAME}`,
      '--catalog',
      `/configs/${SRC_CATALOG_FILENAME}`,
      '--state',
      `/configs/${DEFAULT_STATE_FILE}`,
    ];
    const maxNanoCpus =
      config.src?.dockerOptions?.maxCpus !== undefined ? config.src.dockerOptions.maxCpus * 1e9 : undefined;
    const maxMemory =
      config.src?.dockerOptions?.maxMemory !== undefined ? config.src.dockerOptions.maxMemory * 1024 * 1024 : undefined;
    const createOptions: Docker.ContainerCreateOptions = {
      name: srcContainerName,
      Image: config.src.image,
      Cmd: cmd,
      HostConfig: {
        Binds: [`${tmpDir}:/configs`],
        AutoRemove: true,
        Init: true,
        NanoCpus: maxNanoCpus, // 1e9 nano cpus = 1 cpu
        Memory: maxMemory, // 1024 * 1024 bytes = 1MB
        LogConfig: {
          Type: 'json-file',
          Config: {
            'max-size': config.src?.dockerOptions?.maxLogSize ?? DEFAULT_MAX_LOG_SIZE,
          },
        },
      },
      AttachStdout: true,
      AttachStderr: true,
      Env: [`LOG_LEVEL=${config.logLevel}`],
      platform: 'linux/amd64',
    };

    // Create the Docker container
    const container = await _docker.createContainer(createOptions);

    // Write the container ID to the cidfile
    const cidfilePath = `tmp-${timestamp}-src_cid`;
    writeFileSync(cidfilePath, container.id);

    // create a writable stream to capture the stdout
    let data = '';
    const stdoutStream = new Writable({
      write(chunk, _encoding, callback) {
        data += chunk.toString();
        callback();
      },
    });
    // Attach the stderr to termincal stderr, and stdout to the output stream
    const stream = await container.attach({stream: true, stdout: true, stderr: true});
    container.modem.demuxStream(stream, stdoutStream, process.stderr);

    // Start the container
    await container.start();

    // Wait for the container to finish
    const res = await container.wait();
    logger.debug(res);
    logger.debug(data);

    if (res.StatusCode === 0) {
      logger.info('Source connector ran successfully.');
    } else {
      throw new Error('Failed to run source connector.');
    }

    // return the stdout data
    return data;
  } catch (error: any) {
    throw new Error(`Failed to run source connector: ${error.message ?? JSON.stringify(error)}.`);
  }
}
