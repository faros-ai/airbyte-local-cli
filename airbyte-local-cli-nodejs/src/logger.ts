import pino from 'pino';
import pretty from 'pino-pretty';

// Create a pino logger instance
// export const logger = pino(
//   {
//     level: 'info',
//     transport: undefined as any, // no transport
//   },
//   pino.destination(1), // sync to stdout
// );
export const logger = pino(pretty({colorize: true}));
