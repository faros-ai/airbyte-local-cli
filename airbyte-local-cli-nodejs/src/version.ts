export const CLI_VERSION = '0.0.5';
