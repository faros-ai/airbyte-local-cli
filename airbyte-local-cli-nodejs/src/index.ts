import {parseAndValidateInputs} from './command';
import {checkDockerInstalled, checkSrcConnection, pullDockerImage, runSrcSync} from './docker';
import {AirbyteCliContext} from './types';
import {cleanUp, createTmpDir, loadStateFile, logger, writeConfig} from './utils';

async function main(): Promise<void> {
  const context: AirbyteCliContext = {};
  try {
    // Parse and validate cli arguments
    const cfg = parseAndValidateInputs(process.argv);
    await checkDockerInstalled();

    // Create temporary directory, load state file, write config to files
    context.tmpDir = createTmpDir();
    loadStateFile(context.tmpDir, cfg?.stateFile, cfg?.connectionName);
    writeConfig(context.tmpDir, cfg);

    // Pull source docker image
    if (cfg.srcPull && cfg.src?.image) {
      await pullDockerImage(cfg.src.image);
    }
    // Check source connection
    if (cfg.srcCheckConnection && cfg.src?.image) {
      await checkSrcConnection(context.tmpDir, cfg.src.image);
    }
    // Run airbyte source connector
    if (!cfg.srcInputFile) {
      await runSrcSync(context.tmpDir, cfg);
    }
  } catch (error: any) {
    logger.error(error.message, 'Error');
    cleanUp(context);
    throw error;
  }
}

main().catch((_error) => {
  logger.error('Exit Airbyte CLI with errors.');
  process.exit(1);
});
