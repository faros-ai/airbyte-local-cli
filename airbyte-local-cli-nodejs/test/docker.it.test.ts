import {copyFileSync, readdirSync, readFileSync, rmSync, unlinkSync, writeFileSync} from 'node:fs';
import path from 'node:path';
import {Writable} from 'node:stream';

import {checkSrcConnection, pullDockerImage, runDiscoverCatalog, runDstSync, runSrcSync} from '../src/docker';
import {FarosConfig} from '../src/types';
import {DST_CONFIG_FILENAME, SRC_OUTPUT_DATA_FILE} from '../src/utils';

const defaultConfig: FarosConfig = {
  srcCheckConnection: false,
  dstUseHostNetwork: false,
  srcPull: false,
  dstPull: false,
  fullRefresh: false,
  rawMessages: false,
  keepContainers: false,
  logLevel: 'info',
  debug: false,
  stateFile: undefined,
  connectionName: undefined,
  srcOutputFile: undefined,
  srcInputFile: undefined,
};

beforeAll(async () => {
  await pullDockerImage('farosai/airbyte-example-source');
  await pullDockerImage('farosai/airbyte-faros-graphql-source');
  await pullDockerImage('farosai/airbyte-faros-destination');
});

describe('runDiscoverCatalog', () => {
  it('should success with example source', async () => {
    const res = await runDiscoverCatalog(`${process.cwd()}/test/resources`, 'farosai/airbyte-example-source');

    expect(res).toMatchSnapshot();
  });

  it('should success with graphql source', async () => {
    const res = await runDiscoverCatalog(`${process.cwd()}/test/resources`, 'farosai/airbyte-faros-graphql-source');

    expect(res).toMatchSnapshot();
  });
});

describe('checkSrcConnection', () => {
  it('should success', async () => {
    await expect(
      checkSrcConnection(
        `${process.cwd()}/test/resources`,
        'farosai/airbyte-example-source',
        'faros_airbyte_cli_src_config_chris.json',
      ),
    ).resolves.not.toThrow();
  });

  it('should fail with', async () => {
    await expect(
      checkSrcConnection(
        `${process.cwd()}/test/resources`,
        'farosai/airbyte-example-source',
        'faros_airbyte_cli_src_config_jennie.json',
      ),
    ).rejects.toThrow('Failed to validate source connection: User is not chris.');
  });
});

describe('runSrcSync', () => {
  const testCfg: FarosConfig = {
    ...defaultConfig,
    src: {
      image: 'farosai/airbyte-example-source',
    },
  };

  const testTmpDir = `${process.cwd()}/test/resources/dockerIt_runSrcSync`;

  // remove the intermediate output file
  afterEach(() => {
    rmSync(`${testTmpDir}/${SRC_OUTPUT_DATA_FILE}`, {force: true});
  });

  // Clean up files created by the test
  afterAll(() => {
    const pattern = /.*-src_cid$/;
    const files = readdirSync(process.cwd());
    const matchingFiles = files.filter((file) => pattern.test(file));

    matchingFiles.forEach((file) => {
      const filePath = path.join(process.cwd(), file);
      unlinkSync(filePath);
    });
  });

  it('should success', async () => {
    await expect(runSrcSync(testTmpDir, testCfg)).resolves.not.toThrow();

    // Replace timestamp and version for snapshot comparison
    const output = readFileSync(`${testTmpDir}/${SRC_OUTPUT_DATA_FILE}`, 'utf8');
    const outputWithoutTS = output.split('\n').map((line) => {
      return line
        .replace(/"timestamp":\d+/g, '"timestamp":***')
        .replace(/"sourceVersion":"[\d.]+"/g, '"sourceVersion":***')
        .replace(/Source version: [\d.]+/g, 'Source version: ***');
    });
    expect(outputWithoutTS.join('\n')).toMatchSnapshot();
  });

  // Check stderr message is correctly redirect to process.stderr
  it('should fail', async () => {
    // Capture process.stderr
    let stderrData = '';
    const originalStderrWrite = process.stderr.write;
    const stderrStream = new Writable({
      write(chunk, _encoding, callback) {
        stderrData += chunk.toString();
        callback();
      },
    });
    process.stderr.write = stderrStream.write.bind(stderrStream) as any;

    try {
      await expect(
        runSrcSync(testTmpDir, {
          ...testCfg,
          src: {image: 'farosai/airbyte-faros-graphql-source'},
        }),
      ).rejects.toThrow();
    } finally {
      process.stderr.write = originalStderrWrite;
    }

    expect(stderrData).toContain(`Faros API key was not provided`);
  });
});

describe.only('runDstSync', () => {
  const testCfg: FarosConfig = {
    ...defaultConfig,
    dst: {
      image: 'farosai/airbyte-faros-destination:0.12.5',
    },
    logLevel: 'debug',
    debug: true,
    stateFile: 'testConnectionName__state.json',
  };
  const testTmpDir = `${process.cwd()}/test/resources/dockerIt_runDstSync`;
  const testStateFile = `testConnectionName__state.json`;
  const dstConfigPath = `${testTmpDir}/${DST_CONFIG_FILENAME}`;
  const dstConfigPathTemplate = `${testTmpDir}/${DST_CONFIG_FILENAME}.template`;

  // remove config file that might contain credentials
  afterEach(() => {
    try {
      unlinkSync(dstConfigPath);
    } catch (_error) {
      // ignore
    }
  });

  beforeAll(() => {
    expect(process.env['FAROS_API_KEY']).toBeDefined();
  });

  // Clean up files created by the test
  afterAll(() => {
    const pattern = /.*-dst_cid$/;
    const files = readdirSync(process.cwd());
    const matchingFiles = files.filter((file) => pattern.test(file));

    matchingFiles.forEach((file) => {
      const filePath = path.join(process.cwd(), file);
      unlinkSync(filePath);
    });

    try {
      unlinkSync(testStateFile);
    } catch (_error) {
      // ignore
    }
  });

  it('should success', async () => {
    const dstConfig = JSON.parse(readFileSync(dstConfigPathTemplate, 'utf8'));
    dstConfig.edition_configs.api_key = process.env['FAROS_API_KEY'];
    writeFileSync(dstConfigPath, JSON.stringify(dstConfig, null, 2));

    await expect(runDstSync(testTmpDir, testCfg)).resolves.not.toThrow();

    const stateData = readFileSync(testStateFile, 'utf8');
    expect(JSON.parse(stateData).data).toBeTruthy();
    expect(stateData).toMatchSnapshot();
  }, 60000);

  it('should fail', async () => {
    copyFileSync(dstConfigPathTemplate, dstConfigPath);

    await expect(runDstSync(testTmpDir, testCfg)).rejects.toThrow('Failed to run destination connector');
  }, 60000);
});
