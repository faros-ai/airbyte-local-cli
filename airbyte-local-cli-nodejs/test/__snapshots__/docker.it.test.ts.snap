// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runDiscoverCatalog should success with example source 1`] = `
{
  "streams": [
    {
      "default_cursor_field": [
        "updated_at",
      ],
      "json_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "properties": {
          "fields": {
            "properties": {
              "command": {
                "type": "string",
              },
              "number": {
                "type": "integer",
              },
            },
            "type": "object",
          },
          "more_fields": {
            "items": {
              "properties": {
                "name": {
                  "type": "string",
                },
                "nested": {
                  "properties": {
                    "value": {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "value": {
                  "type": [
                    "string",
                    "integer",
                  ],
                },
              },
              "type": "object",
            },
            "type": [
              "null",
              "array",
            ],
          },
          "source": {
            "type": "string",
          },
          "uid": {
            "type": "string",
          },
          "updated_at": {
            "type": "integer",
          },
        },
        "type": "object",
      },
      "name": "builds",
      "source_defined_cursor": true,
      "source_defined_primary_key": [
        [
          "uid",
        ],
        [
          "source",
        ],
      ],
      "supported_sync_modes": [
        "full_refresh",
        "incremental",
      ],
    },
  ],
}
`;

exports[`runDiscoverCatalog should success with graphql source 1`] = `
{
  "streams": [
    {
      "default_cursor_field": [],
      "json_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "properties": {
          "message": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "name": "faros_graph",
      "source_defined_cursor": true,
      "supported_sync_modes": [
        "full_refresh",
        "incremental",
      ],
    },
  ],
}
`;

exports[`runDstSync should success 1`] = `"{"format":"base64/gzip","data":"H4sIAAAAAAAAA6tWSkssyi+OTy9KLMhQsqpWyi9Kjw9JTcwFsYtS04pSizNSUxxLfDNzcjKLlawMzY0NzczNLI1NTC3NamtrAc6d5gdAAAAA"}"`;

exports[`runDstSync should success with srcInputFile 1`] = `"{"format":"base64/gzip","data":"H4sIAAAAAAAAA6tWSkssyi+OTy9KLMhQsqpWyi9Kjw9JTcwFsYtS04pSizNSUxxLfDNzcjKLlawMzY0NzczNLI1NTC3NamtrAc6d5gdAAAAA"}"`;

exports[`runSpec should success with airbyte image 1`] = `
{
  "spec": {
    "connectionSpecification": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "properties": {
        "accept_terms": {
          "default": false,
          "description": "You must agree to the Databricks JDBC Driver <a href="https://databricks.com/jdbc-odbc-driver-license">Terms & Conditions</a> to use this connector.",
          "order": 1,
          "title": "Agree to the Databricks JDBC Driver Terms & Conditions",
          "type": "boolean",
        },
        "authentication": {
          "default": "OAUTH",
          "description": "Authentication mechanism for Staging files and running queries",
          "oneOf": [
            {
              "properties": {
                "auth_type": {
                  "const": "OAUTH",
                  "order": 0,
                  "type": "string",
                },
                "client_id": {
                  "order": 1,
                  "type": "string",
                },
                "secret": {
                  "airbyte_secret": true,
                  "order": 2,
                  "type": "string",
                },
              },
              "required": [
                "auth_type",
                "client_id",
                "secret",
              ],
              "title": "OAuth2 (Recommended)",
            },
            {
              "properties": {
                "auth_type": {
                  "const": "BASIC",
                  "order": 0,
                  "type": "string",
                },
                "personal_access_token": {
                  "airbyte_secret": true,
                  "order": 1,
                  "type": "string",
                },
              },
              "required": [
                "auth_type",
                "personal_access_token",
              ],
              "title": "Personal Access Token",
            },
          ],
          "order": 8,
          "title": "Authentication",
          "type": "object",
        },
        "database": {
          "description": "The name of the unity catalog for the database",
          "order": 5,
          "title": "Databricks Unity Catalog Name",
          "type": "string",
        },
        "hostname": {
          "description": "Databricks Cluster Server Hostname.",
          "examples": [
            "abc-12345678-wxyz.cloud.databricks.com",
          ],
          "order": 2,
          "title": "Server Hostname",
          "type": "string",
        },
        "http_path": {
          "description": "Databricks Cluster HTTP Path.",
          "examples": [
            "sql/1.0/warehouses/0000-1111111-abcd90",
          ],
          "order": 3,
          "title": "HTTP Path",
          "type": "string",
        },
        "port": {
          "default": "443",
          "description": "Databricks Cluster Port.",
          "examples": [
            "443",
          ],
          "order": 4,
          "title": "Port",
          "type": "string",
        },
        "purge_staging_data": {
          "default": true,
          "description": "Default to 'true'. Switch it to 'false' for debugging purpose.",
          "order": 9,
          "title": "Purge Staging Files and Tables",
          "type": "boolean",
        },
        "raw_schema_override": {
          "default": "airbyte_internal",
          "description": "The schema to write raw tables into (default: airbyte_internal)",
          "order": 10,
          "title": "Raw Table Schema Name",
          "type": "string",
        },
        "schema": {
          "default": "default",
          "description": "The default schema tables are written. If not specified otherwise, the "default" will be used.",
          "examples": [
            "default",
          ],
          "order": 6,
          "title": "Default Schema",
          "type": "string",
        },
      },
      "required": [
        "accept_terms",
        "hostname",
        "http_path",
        "database",
        "authentication",
      ],
      "title": "Databricks Lakehouse Destination Spec",
      "type": "object",
    },
    "documentationUrl": "https://docs.airbyte.com/integrations/destinations/databricks",
    "supported_destination_sync_modes": [
      "overwrite",
      "append",
      "append_dedup",
    ],
    "supportsDBT": false,
    "supportsIncremental": true,
    "supportsNormalization": false,
  },
  "type": "SPEC",
}
`;

exports[`runSpec should success with faros image 1`] = `
{
  "spec": {
    "connectionSpecification": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "api_key": {
          "airbyte_secret": true,
          "description": "The Faros API key to use to access the API.",
          "order": 1,
          "title": "API Key",
          "type": "string",
        },
        "api_url": {
          "default": "https://prod.api.faros.ai",
          "description": "The Faros API URL.",
          "examples": [
            "https://prod.api.faros.ai",
          ],
          "order": 0,
          "title": "API URL",
          "type": "string",
        },
        "bucket_id": {
          "default": 1,
          "description": "Bucket number for this source to determine which portion of models to pull. Use it when distributing the load between multiple sources. Cannot be used in combination with 'GraphQL query'. Please note that V1 graphs do not support opening multiple revisions at once, so only use buckets if writing to a V2 graph on the destination, or not writing at all.",
          "order": 10,
          "title": "Bucket Number",
          "type": "integer",
        },
        "bucket_total": {
          "default": 1,
          "description": "Total number of buckets to distribute models across. Use it when distributing the load between multiple sources. Cannot be used in combination with 'GraphQL query'. Please note that V1 graphs do not support opening multiple revisions at once, so only use buckets if writing to a V2 graph on the destination, or not writing at all.",
          "order": 11,
          "title": "Total Number of Buckets",
          "type": "integer",
        },
        "debug": {
          "default": false,
          "description": "Enable debug mode",
          "order": 1000,
          "title": "Debug",
          "type": "boolean",
        },
        "faros_source_id": {
          "description": "The ID of the source (aka account)",
          "order": 1001,
          "title": "The source ID",
          "type": "string",
        },
        "graph": {
          "default": "default",
          "description": "The graph name.",
          "order": 2,
          "title": "Graph name",
          "type": "string",
        },
        "models_filter": {
          "description": "Only query the following models and skip everything else.",
          "items": {
            "type": "string",
          },
          "order": 5,
          "title": "Models filter",
          "type": "array",
        },
        "page_size": {
          "default": 100,
          "description": "Page size to use when paginating through query results.",
          "order": 6,
          "title": "Page Size",
          "type": "integer",
        },
        "query": {
          "description": "The query to execute.",
          "multiline": true,
          "order": 4,
          "title": "GraphQL query",
          "type": "string",
        },
        "result_model": {
          "default": "Nested",
          "description": "Where to place each of the query results in their corresponding output records. E.g Nested - { vcs { pullRequests { nodes: [<record>] } } }, Flat - { vcs_PullRequest: <record> }",
          "enum": [
            "Nested",
            "Flat",
          ],
          "order": 7,
          "title": "Result model",
          "type": "string",
        },
      },
      "required": [
        "api_key",
        "graph",
      ],
      "title": "GraphQL Spec",
      "type": "object",
    },
    "documentationUrl": "https://docs.faros.ai",
  },
  "type": "SPEC",
}
`;

exports[`runSrcSync should success 1`] = `
"{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","redactedConfig":{"user":"chris"},"sourceType":"example","sourceVersion":***}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","sourceStatus":{"status":"SUCCESS"}}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","logs":[{"timestamp":***,"message":{"level":30,"msg":"Source version: ***"}},{"timestamp":***,"message":{"level":30,"msg":"Config: {\\"user\\":\\"chris\\"}"}},{"timestamp":***,"message":{"level":30,"msg":"Catalog: {}"}},{"timestamp":***,"message":{"level":30,"msg":"State: {}"}},{"timestamp":***,"message":{"level":30,"msg":"Syncing ExampleSource"}},{"timestamp":***,"message":{"level":30,"msg":"Finished syncing ExampleSource"}}]}
"
`;

exports[`runSrcSync should success with specified output file 1`] = `
"{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","redactedConfig":{"user":"chris"},"sourceType":"example","sourceVersion":***}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","sourceStatus":{"status":"SUCCESS"}}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","logs":[{"timestamp":***,"message":{"level":30,"msg":"Source version: ***"}},{"timestamp":***,"message":{"level":30,"msg":"Config: {\\"user\\":\\"chris\\"}"}},{"timestamp":***,"message":{"level":30,"msg":"Catalog: {}"}},{"timestamp":***,"message":{"level":30,"msg":"State: {}"}},{"timestamp":***,"message":{"level":30,"msg":"Syncing ExampleSource"}},{"timestamp":***,"message":{"level":30,"msg":"Finished syncing ExampleSource"}}]}
"
`;

exports[`runSrcSync should wait for write to complete 1`] = `
"{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","redactedConfig":{"user":"chris"},"sourceType":"example","sourceVersion":***}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","sourceStatus":{"status":"SUCCESS"}}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","logs":[{"timestamp":***,"message":{"level":30,"msg":"Source version: ***"}},{"timestamp":***,"message":{"level":30,"msg":"Config: {\\"user\\":\\"chris\\"}"}},{"timestamp":***,"message":{"level":30,"msg":"Catalog: {}"}},{"timestamp":***,"message":{"level":30,"msg":"State: {}"}},{"timestamp":***,"message":{"level":30,"msg":"Syncing ExampleSource"}},{"timestamp":***,"message":{"level":30,"msg":"Finished syncing ExampleSource"}}]}
"
`;

exports[`runWizard should success with airbyte image 1`] = `
{
  "accept_terms": false,
  "database": "<UPDATE_THIS_VALUE>",
  "hostname": "<UPDATE_THIS_VALUE>",
  "http_path": "<UPDATE_THIS_VALUE>",
}
`;

exports[`runWizard should success with faros image 1`] = `
{
  "api_key": "<UPDATE_YOUR_CREDENTIAL>",
  "graph": "default",
}
`;
